File: templates/figures/python/plot_optimization_convergence.py
from __future__ import annotations

import argparse

import matplotlib.pyplot as plt

from recipe_common import COLORS, configure_style, direct_label, export_triplet, label, load_verified_csv, new_figure, numeric, style_axis


def main() -> int:
    parser = argparse.ArgumentParser(description="Plot multi-run optimization convergence with a matched baseline.")
    parser.add_argument("--input", required=True)
    parser.add_argument("--output-dir", required=True)
    parser.add_argument("--stem", default="optimization-convergence")
    parser.add_argument("--objective-label", required=True)
    parser.add_argument("--objective-unit", required=True)
    args = parser.parse_args()

    configure_style()
    required = ["iteration", "run", "objective", "baseline_objective"]
    frame = load_verified_csv(args.input, required)
    numeric(frame, required)
    if frame.duplicated(["iteration", "run"]).any():
        raise ValueError("Each iteration-run pair must be unique")

    fig, ax = new_figure(height_mm=100)
    for column, color, linestyle, marker, series_label in (
        ("baseline_objective", COLORS["baseline"], "--", "s", "Baseline"),
        ("objective", COLORS["primary"], "-", "o", "Main model"),
    ):
        grouped = frame.groupby("iteration")[column]
        mean = grouped.mean()
        low = grouped.quantile(0.05)
        high = grouped.quantile(0.95)
        ax.fill_between(mean.index, low, high, color=color, alpha=0.14, linewidth=0)
        ax.plot(mean.index, mean, color=color, linestyle=linestyle, marker=marker, markevery=max(1, len(mean) // 10), label=f"_{series_label}")
        role = "baseline" if column == "baseline_objective" else "main_model"
        direct_label(ax, float(mean.index[-1]), float(mean.iloc[-1]), series_label, role=role, offset_points=(6, -5 if role == "baseline" else 5))
    ax.set(xlabel="Iteration", ylabel=label(args.objective_label, args.objective_unit))
    style_axis(ax)
    export_triplet(fig, args.output_dir, args.stem)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: templates/figures/matlab/plot_optimization_convergence.m
function outputs = plot_optimization_convergence(inputCsv, outputDir, stem, objectiveLabel, objectiveUnit)
% Plot an observed solver trace against a comparable baseline.
arguments
    inputCsv (1,1) string
    outputDir (1,1) string
    stem (1,1) string = "optimization-convergence"
    objectiveLabel (1,1) string = "Objective"
    objectiveUnit (1,1) string = "dimensionless"
end
T = readtable(inputCsv, 'TextType', 'string');
mm_require_columns(T, {'iteration', 'objective', 'baseline_objective'});
iteration = mm_num(T, 'iteration');
objective = mm_num(T, 'objective');
baseline = mm_num(T, 'baseline_objective');
[iteration, order] = sort(iteration);
objective = objective(order);
baseline = baseline(order);
if any(diff(iteration) < 0)
    error('mathmodeling:IterationOrder', 'Iteration values must be sortable.');
end

fig = mm_publication_figure(15, 8.2);
ax = axes(fig);
plot(ax, iteration, objective, 'Color', [59 111 182] / 255, 'LineWidth', 1.5, 'DisplayName', 'Solver trace');
hold(ax, 'on');
plot(ax, iteration, baseline, '--', 'Color', [217 119 50] / 255, 'LineWidth', 1.2, 'DisplayName', 'Baseline');
hold(ax, 'off');
xlabel(ax, 'Iteration');
if objectiveUnit == "dimensionless"
    ylabel(ax, objectiveLabel);
else
    ylabel(ax, objectiveLabel + " (" + objectiveUnit + ")");
end
legend(ax, 'Location', 'best');
grid(ax, 'on');
ax.GridAlpha = 0.18;
outputs = mm_export_triplet(fig, outputDir, stem);
close(fig);
end
